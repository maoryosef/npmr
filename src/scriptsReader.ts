import { join } from 'path';
import { cwd } from 'process';
import { readFileSync, existsSync } from 'fs';

interface PackageJson {
    scripts: Record<string, string>;
}

function readPackageJson(): PackageJson | null{
    const fileName = join(cwd(), 'package.json');
    
    if (existsSync(fileName)) {
        return JSON.parse(readFileSync(fileName, 'utf-8'));
    }
    
    return null;
}

export interface PackageJsonScript {
    name: string;
    cmd: string;
}

export const readScripts = (): PackageJsonScript[] =>
    Object.entries(readPackageJson()?.scripts || {})
        .reduce<PackageJsonScript[]>((acc, [name, cmd]) => [...acc, { name, cmd }], [])
