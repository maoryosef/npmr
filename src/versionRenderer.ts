import pkg from '../package.json'

export const render = (): void => {
    console.log(`
=============================================
    npmr version: ${pkg.version}             
=============================================
`)
}
