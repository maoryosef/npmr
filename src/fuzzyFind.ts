import fuzzysort from 'fuzzysort';
import { PackageJsonScript } from './scriptsReader';

const KEYS = ['cmd', 'name']
const HIGHLIGHT_SPLIT_CHAR = '\x00'
export interface Highlight {
  value: string;
  highlighted: boolean;
}
export interface SearchResult {
    value: PackageJsonScript;
    score: number;
    highlights: {
      cmd?: Highlight[];
      name?: Highlight[];
    }
}

const packageJsonScriptToSearchResult = (value: PackageJsonScript) : SearchResult =>
  ({value, score: 0, highlights: {}})

const toHighlights = (highlightedString: string | null) : Highlight[] | null => {
    if (!highlightedString) {
      return null;
    }
    
    const highlights: Highlight[] = []
    let highlighted = false;
    let value = ''    
    
    for (const c of highlightedString) {
      if (c === HIGHLIGHT_SPLIT_CHAR) {
        highlights.push({
          highlighted,
          value
        })
        
        value = ''
        highlighted = !highlighted
      } else {
        value += c
      }
    }
    
    if (value.length > 0) {
      highlights.push({
        highlighted: false,
        value
      })
    }
    
    return highlights
}

const getHighlights = (result: Fuzzysort.Result[], idx: number): Highlight[] | null =>
  toHighlights(fuzzysort.highlight(result[idx], HIGHLIGHT_SPLIT_CHAR, HIGHLIGHT_SPLIT_CHAR))


const fuzzySortResultToSearchResult = (result: Fuzzysort.KeyResult<PackageJsonScript>) : SearchResult =>
 ({
  value: result.obj,
  score: result.score,
  highlights: KEYS.reduce((obj, key, idx) =>
    ({ ...obj, [key]: getHighlights(result as unknown as Fuzzysort.Result[], idx) }), {}),
})

const doFuzzySort = (list: Array<PackageJsonScript>, query: string): SearchResult[] =>
  fuzzysort
    .go(query, list, {
      keys: KEYS,
      limit: 16, // Don't return more results than this (lower is faster)
      allowTypo: true, // (false is faster)
    })
    // @ts-ignore
    .sort((a, b) => a.score < b.score)
    .map(fuzzySortResultToSearchResult)

export default (list: Array<PackageJsonScript>, query: string = ''): SearchResult[] => 
    query.trim().length > 0 ? 
      doFuzzySort(list, query) :
      list.map(packageJsonScriptToSearchResult)