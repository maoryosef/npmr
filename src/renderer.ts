import chalk from 'chalk';
import figures from 'figures';

import { PackageJsonScript } from './scriptsReader';

export function renderValue(value: string, cursorPos: number): string {
    const prefix = chalk.cyan(`Search task${figures.pointer} `)
    
    if (value === '') {
        return prefix + value + chalk.dim.gray('Start typing to search...');
    }
    
    if (cursorPos === value.length) {
        return prefix + value + chalk.inverse(' ');
    }

    const firstChunk = value.slice(0, cursorPos);
    const corsurChar = value.slice(cursorPos, cursorPos + 1);
    const secondChunk = value.slice(cursorPos + 1);

    return prefix + firstChunk + chalk.inverse(corsurChar) + secondChunk;
}

export function renderScripts(scripts: PackageJsonScript[], selectedIdx: number): string {
    return scripts.map((script, idx) =>
        selectedIdx === idx ? 
            `${figures.play} ${chalk.bold(script.name)}\t ${chalk.white(script.cmd)}` :
            `  ${chalk.dim(script.name)}\t ${chalk.dim(script.cmd)}`
    ).join('\n');
}