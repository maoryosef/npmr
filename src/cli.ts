import logUpdate from 'log-update';
import InputHandler, { InputKey } from './input/inputHandler';
import * as terminal from './utils/terminal';
import {clamp} from './utils/math';
import * as renderer from './renderer';
import * as reader from './scriptsReader';
import { spawnSync } from 'child_process';
import fuzzyFind from './fuzzyFind';

const input = new InputHandler({ 
    stdin: process.stdin
})

terminal.clear();

const scripts = reader.readScripts();
let results = fuzzyFind(scripts)
let selectedIdx = 0;

function render(handler: InputHandler) {
    logUpdate([
        renderer.renderSearchString(handler.value, handler.cursorPos),
        renderer.renderScripts(results, selectedIdx),
    ].join('\n')
    )
}

render(input);

input.on('change', (handler, key) => {
    switch (key) {
        case InputKey.DOWN: selectedIdx++; break;
        case InputKey.UP: selectedIdx--; break;
    }
    
    results = fuzzyFind(scripts, handler.value);
    selectedIdx = clamp(selectedIdx, 0, results.length - 1);

    render(handler);
})

input.on('select', () => {
    terminal.clear();
    input.end();
    spawnSync('yarn', [scripts[selectedIdx].cmd], {stdio: 'inherit', shell: true});

    process.exit();
})

input.on('exit', () => {
    terminal.clear();
    input.end();
    process.exit();
})