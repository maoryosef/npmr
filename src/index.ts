import logUpdate from 'log-update';
import InputHandler, { InputKey } from './input/inputHandler';
import * as terminal from './utils/terminal';
import * as renderer from './renderer';
import * as reader from './scriptsReader';
import { spawnSync } from 'child_process';

const input = new InputHandler({ 
    stdin: process.stdin
})

terminal.clear();

const scripts = reader.readScripts();

let selectedIdx = 0;

function render(handler: InputHandler) {
    logUpdate([
        renderer.renderValue(handler.value, handler.cursorPos),
        renderer.renderScripts(scripts, selectedIdx),
    ].join('\n')
    )
}

render(input);

function clamp(value: number, min: number, max: number) {
    if (value < min) {
        return min;
    }
    
    if (value > max) {
        return max;
    }
    
    return value;
}

input.on('change', (handler, key) => {
    switch (key) {
        case InputKey.DOWN: selectedIdx++; break;
        case InputKey.UP: selectedIdx--; break;
    }
    
    selectedIdx = clamp(selectedIdx, 0, scripts.length - 1);

    render(handler);
})

input.on('select', () => {
    terminal.clear();
    input.end();
    spawnSync('yarn', [scripts[selectedIdx].cmd], {stdio: 'inherit', shell: true});

    process.exit();
})

input.on('exit', () => {
    terminal.clear();
    input.end();
    process.exit();
})