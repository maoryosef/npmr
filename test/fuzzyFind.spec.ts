import {setFixture, beforeAndAfter} from './fixtures'
import {PackageJsonScript, readScripts} from '../src/scriptsReader'
import fuzzyFind from '../src/fuzzyFind';

beforeAndAfter()

describe('fuzzyFind', () => {
    describe('when query is empty', () => {
        test('should return all scripts', () => {
            setFixture('simplePackage')
            const scripts = readScripts()
            
            expect(fuzzyFind(scripts)).toMatchSnapshot()
        })
    })

    describe('when query is not empty', () => {
        test.each(['webpack', 'tsc'])('should return matching scripts for query "%s"', query => {
            setFixture('simplePackage')
            const scripts = readScripts()
            
            expect(fuzzyFind(scripts, query)).toMatchSnapshot()
        })
    })
    
    describe('highilights', () => {
        test('should return highlighted object', () => {
            const scripts: PackageJsonScript[] = [{
                cmd: 'tsc --watch',
                name: 'build'
            }]
            
            const [result] = fuzzyFind(scripts, 'sc')
            
            expect(result.highlights.name).toBe(null)
            expect(result.highlights.cmd).toEqual([
                { highlighted: false, value: 't'},
                { highlighted: true, value: 'sc'},
                { highlighted: false, value: ' --watch'}
            ])
        })

        describe('when highlight contains multiple parts', () => {
            test('should return highlighted object', () => {
                const scripts: PackageJsonScript[] = [{
                    cmd: 'jest',
                    name: 'test'
                }]
                
                const [result] = fuzzyFind(scripts, 'ts')

                expect(result.highlights.name).toEqual([
                    { highlighted: false, value: ''},
                    { highlighted: true, value: 't'},
                    { highlighted: false, value: 'e'},
                    { highlighted: true, value: 's'},
                    { highlighted: false, value: 't'},
                ])
            })
        })
    })
})