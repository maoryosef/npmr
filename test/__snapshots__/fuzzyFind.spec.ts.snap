// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fuzzyFind when query is empty should return all scripts 1`] = `
Array [
  Object {
    "highlights": Object {},
    "score": 0,
    "value": Object {
      "cmd": "tsc",
      "name": "build",
    },
  },
  Object {
    "highlights": Object {},
    "score": 0,
    "value": Object {
      "cmd": "webpack",
      "name": "build:webpack",
    },
  },
  Object {
    "highlights": Object {},
    "score": 0,
    "value": Object {
      "cmd": "tsc --watch",
      "name": "watch",
    },
  },
]
`;

exports[`fuzzyFind when query is not empty should return matching scripts for query "tsc" 1`] = `
Array [
  Object {
    "highlights": Object {
      "cmd": Array [
        Object {
          "highlighted": false,
          "value": "",
        },
        Object {
          "highlighted": true,
          "value": "tsc",
        },
      ],
      "name": null,
    },
    "score": 0,
    "value": Object {
      "cmd": "tsc",
      "name": "build",
    },
  },
  Object {
    "highlights": Object {
      "cmd": Array [
        Object {
          "highlighted": false,
          "value": "",
        },
        Object {
          "highlighted": true,
          "value": "tsc",
        },
        Object {
          "highlighted": false,
          "value": " --watch",
        },
      ],
      "name": null,
    },
    "score": -8,
    "value": Object {
      "cmd": "tsc --watch",
      "name": "watch",
    },
  },
]
`;

exports[`fuzzyFind when query is not empty should return matching scripts for query "webpack" 1`] = `
Array [
  Object {
    "highlights": Object {
      "cmd": Array [
        Object {
          "highlighted": false,
          "value": "",
        },
        Object {
          "highlighted": true,
          "value": "webpack",
        },
      ],
      "name": Array [
        Object {
          "highlighted": false,
          "value": "build:",
        },
        Object {
          "highlighted": true,
          "value": "webpack",
        },
      ],
    },
    "score": 0,
    "value": Object {
      "cmd": "webpack",
      "name": "build:webpack",
    },
  },
]
`;
